modules/math/shared/src/main/scala/gsp/math/skycalc/ImprovedSkyCalc.scala

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gsp.math.skycalc
+
+import cats.Eq
+import cats.implicits._
+import gsp.math.Angle
+import gsp.math.Coordinates
+import java.time.Instant
+import java.time.ZonedDateTime
+
+/**
+  * Improved version of SkyCalc that supports lunar calculations. All instance stuff is here;
+  * the trait is exclusively static stuff.
+  * @author brighton, rnorris
+  */
+final class ImprovedSkyCalc(siteLatitude: Angle, siteLongitude: Angle, siteAlt: Int)
+    extends ImprovedSkyCalcMethods {
+
+  // Site parameters
+  private val hoursLongitude  = -siteLongitude.toDoubleDegrees / 15.0
+  private val degreesLatitude = siteLatitude.toDoubleDegrees
+  private val siteAltitude    = siteAlt.toDouble
+
+  // calculated results
+  private var altitude                             = .0
+  private var hourAngle                            = .0
+  private var azimuth                              = .0
+  private var parallacticAngle                     = .0
+  private var airmass                              = .0
+  private var lunarSkyBrightness: java.lang.Double = .0
+  private var lunarDistance                        = .0
+  private var lunarIlluminatedFraction             = .0
+  private var totalSkyBrightness                   = .0
+  private var lunarPhaseAngle                      = .0
+  private var sunAltitude                          = .0
+  private var lunarElevation                       = .0
+
+  // caching for calculate()
+  private var cachedCoordinates: Option[Coordinates] = none
+  private var cachedInstant: Option[Instant]         = none
+  private var cachedCalculateMoon: Option[Boolean]   = none
+
+  private implicit val eqInstant: Eq[Instant] = Eq.by(i => (i.getEpochSecond, i.getNano))
+
+  def calculate(
+    coords:        Coordinates,
+    instant:       Instant,
+    calculateMoon: Boolean
+  ): Unit = { // Early exit if the parameters haven't changed.
```

Comment:
Eventually it would be nice to return a struct here rather than side-effecting, and add an LRU cache. But for now this is ok I think.

Reply:
That sounds good and it should be straightforward. I'll try to take it a notch further and encapsulate init/calc/get in one static function and expose only that.

BTW, I think it would be nice to have a `Location` class that encapsulates (latitude, longitude, altitude) and use that in the input. In that case we could make `Site` in `gsp-core` extend this class. I accept suggestions for a more accurate name other than `Location`.

Reply:
About the cache: I'm not sure it would be useful unless we specify a really big cache. For example, for a night, we compute a point for every minute, which is in the order of 600 points. Caching would be useful if the user switches back and forth between nights/targets/sites, so it would make sense if we store data for a bunch of charts. Do we want, say, a 6000 entry cache?

Also, maybe in that case we want to cache at a higher level, having one cache entry store the entire plot data (QPT does that).

I'm only thinking in terms of plots here, I don't know yet if this code is used in other scenarios.

WDYT?

Reply:
Do we really need a cache? If we were I'd make it external to the code here so we can use it only if needed

Reply:
That's also fine, but we do want a way to do it because these calculations are expensive.

Reply:
I removed the cache for now, mainly because it didn't seem thread safe. Let's implement it on callers as needed.

Reply:
I'd appreciate another look at this. I added a case class to return the results with typed accessors, but I'm not 100% sure I got them right. I had to keep the raw double values there in order to compare them with the output from the OCS version, but they are only visible from the package.